pkg/ingester/wal.go

```diff
@@ -0,0 +1,726 @@
+	w.wait.Wait()
+
+	w.wal.Close()
+}
+
+func (w *walWrapper) Log(record *Record) error {
+	if record == nil {
+		return nil
+	}
+	buf, err := proto.Marshal(record)
+	if err != nil {
+		return err
+	}
+	return w.wal.Log(buf)
+}
+
+func (w *walWrapper) run() {
+	defer w.wait.Done()
+
+	if !w.cfg.checkpointEnabled {
+		return
+	}
+
+	ticker := time.NewTicker(w.cfg.checkpointDuration)
+	defer ticker.Stop()
+
+	for !w.isStopped() {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			level.Info(util.Logger).Log("msg", "starting checkpoint")
+			if err := w.checkpoint(); err != nil {
+				level.Error(util.Logger).Log("msg", "error checkpointing series", "err", err)
+				continue
+			}
+			elapsed := time.Since(start)
+			level.Info(util.Logger).Log("msg", "checkpoint done", "time", elapsed.String())
+		case <-w.quit:
+			if err := w.checkpoint(); err != nil {
+				level.Error(util.Logger).Log("msg", "error checkpointing series during shutdown", "err", err)
+			}
+			return
+		}
+	}
+}
+
+func (w *walWrapper) isStopped() bool {
+	select {
+	case <-w.quit:
+		return true
+	default:
+		return false
+	}
+}
+
+const checkpointPrefix = "checkpoint."
+
+func (w *walWrapper) checkpoint() (err error) {
+	if !w.cfg.checkpointEnabled {
+		return nil
+	}
+
+	w.checkpointCreationTotal.Inc()
+	defer func() {
+		if err != nil {
+			w.checkpointCreationFail.Inc()
+		}
+	}()
+
+	if w.getUserStates == nil {
+		return errors.New("function to get user states not initialised")
+	}
+
+	_, lastSegment, err := w.wal.Segments()
+	if err != nil {
+		return err
+	}
+
+	_, lastCh, err := lastCheckpoint(w.wal.Dir())
+	if err != nil {
+		return err
+	}
+
+	newIdx := lastCh + 1
+
+	cpdir := filepath.Join(w.wal.Dir(), fmt.Sprintf(checkpointPrefix+"%06d", newIdx))
+	level.Info(util.Logger).Log("msg", "attempting checkpoint for", "dir", cpdir)
+	cpdirtmp := cpdir + ".tmp"
+
+	if err := os.MkdirAll(cpdirtmp, 0777); err != nil {
+		return errors.Wrap(err, "create checkpoint dir")
+	}
+	cp, err := wal.New(nil, nil, cpdirtmp, true)
+	if err != nil {
+		return errors.Wrap(err, "open checkpoint")
+	}
+	defer func() {
+		cp.Close()
+		os.RemoveAll(cpdirtmp)
+	}()
+
+	var wireChunkBuf []client.Chunk
+	for userID, state := range w.getUserStates() {
+		for pair := range state.fpToSeries.iter() {
+			state.fpLocker.Lock(pair.fp)
+			wireChunkBuf, err = w.checkpointSeries(cp, userID, pair.fp, pair.series, wireChunkBuf)
+			state.fpLocker.Unlock(pair.fp)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := cp.Close(); err != nil {
+		return errors.Wrap(err, "close checkpoint")
+	}
+	if err := fileutil.Replace(cpdirtmp, cpdir); err != nil {
+		return errors.Wrap(err, "rename checkpoint directory")
+	}
+
+	// The last segment might still have been active during the checkpointing,
+	// hence delete only the segments before that.
+	if err := w.wal.Truncate(lastSegment - 1); err != nil {
+		return err
+	}
+
+	if lastCh >= 0 {
+		if err := w.deleteCheckpoints(lastCh); err != nil {
```

Comment:
Will add some comments. But looking at this, there might be some problems in the way we are naming the checkpoints now. In Prometheus we name it based on the last WAL segment so that we can continue replaying from there. Here we rely on the WAL truncation so that we dont replay the old data, and we only increment the checkpoint number. I think I will change it to the Prometheus way, that way both of them will be just logs.